apps/src/weblab/brambleHost.js

```diff
@@ -65,6 +68,100 @@ function putFilesInBramble(Bramble, sources, callback) {
   });
 }
 
+function removeAllAssetsInBramble(Bramble, callback) {
+  const fs = bramble_.getFileSystem();
+  const sh = new fs.Shell();
+  const Path = bramble_.Filer.Path;
+
+  // enumerate files in the file system off the project root
+  sh.ls(projectRoot, function (err, entries) {
+    if (err) {
+      throw err;
+    }
+    // async-chained enumeration: remove the file if the attributes indicate
+    // that it is an asset
+    function removeEntryIfAsset(i, callback) {
+      if (i < entries.length) {
+        const entry = entries[i];
+        const path = Path.join(projectRoot, entry.path);
+        fs.getxattr(path, 'asset', (err, assetAttrValue) => {
+          if (err && err.name !== 'ENOATTR') {
+            throw err;
+          }
+          if (assetAttrValue) {
+            sh.rm(path, (err) => {
+              if (err) {
+                throw err;
+              }
+              // File removed, move to next file entry:
+              removeEntryIfAsset(i + 1, callback);
+            });
+            return;
+          }
+          // Not an asset file, move to next file entry:
+          removeEntryIfAsset(i + 1, callback);
+        });
+      } else {
+        // end of list, call completion callback
+        callback();
+      }
+    }
+    // start an async-chained enumeration through the file list
+    removeEntryIfAsset(0, callback);
+  });
+}
+
+function syncAssetsWithBramble(Bramble, assets, callback) {
+  const fs = bramble_.getFileSystem();
+  const sh = new fs.Shell();
+  const Path = bramble_.Filer.Path;
+  const Buffer = bramble_.Filer.Buffer;
+  assets = assets || [];
+
+  function requestAssetAndWrite(asset, callback) {
+    // read the data
+    $.ajax(asset.url, {
+      dataType: 'binary',
+      responseType: 'arraybuffer'
+    }).done((data) => {
+      var path = Path.join(projectRoot, asset.name);
+      // write the data
+      fs.writeFile(path, new Buffer(data), { encoding: null }, (err) => {
+        if (err) {
+          throw err;
+        }
+        // mark the extended attribute 'asset' as true
+        fs.setxattr(path, 'asset', true, (err) => {
+          if (err) {
+            throw err;
+          }
+          // call completion callback
+          callback();
```

Comment:
This function has a few ways it can fail to reach this point and call the callback, which would result in `writeAssetFile` stopping in the middle of the list and never calling _its_ callback either.  I'm not sure how far back it goes, but this seems likely not to recover gracefully from an error.  Switching to a `Promise` solution or using a node-style `onError` callback would help with this.